toolchain/check/eval.cpp

```diff
@@ -520,112 +546,81 @@ static auto ReplaceFieldWithConstantValue(EvalContext& eval_context,
-          typename... EachFieldIdT>
-static auto RebuildIfFieldsAreConstantImpl(
-    EvalContext& eval_context, SemIR::Inst inst, ValidateFn validate_fn,
-    TransformFn transform_fn, EachFieldIdT InstT::*... each_field_id)
-    -> SemIR::ConstantId {
-  // Build a constant instruction by replacing each non-constant operand with
-  // its constant value.
-  auto typed_inst = inst.As<InstT>();
-  Phase phase = Phase::Concrete;
-  if ((ReplaceFieldWithConstantValue(eval_context, &typed_inst, each_field_id,
-                                     &phase) &&
-       ...)) {
-    if (phase == Phase::UnknownDueToError || !validate_fn(typed_inst)) {
-      return SemIR::ErrorInst::SingletonConstantId;
-    }
-    return MakeConstantResult(eval_context.context(), transform_fn(typed_inst),
-                              phase);
-  }
-  return MakeNonConstantResult(phase);
-}
-
-// Same as above but with an identity transform function.
-template <typename InstT, typename ValidateFn, typename... EachFieldIdT>
-static auto RebuildAndValidateIfFieldsAreConstant(
-    EvalContext& eval_context, SemIR::Inst inst, ValidateFn validate_fn,
-    EachFieldIdT InstT::*... each_field_id) -> SemIR::ConstantId {
-  return RebuildIfFieldsAreConstantImpl(eval_context, inst, validate_fn,
-                                        std::identity{}, each_field_id...);
-}
+// Function template that can be called with an argument of type `T`. Used below
+// to detect which overloads of `GetConstantValue` exist.
+template <typename T>
+static void Accept(T /*arg*/) {}
+
+// Determines whether a `GetConstantValue` overload exists for a given ID type.
+// Note that we do not check whether `GetConstantValue` is *callable* with a
+// given ID type, because that would use the `InstId` overload for
+// `AbsoluteInstId` and similar wrapper types, which should be left alone.
+template <typename IdT>
+static constexpr bool HasGetConstantValueOverload = requires {
+  Accept<auto (*)(EvalContext&, IdT, Phase*)->IdT>(GetConstantValue);
+};
 
-// Same as above but with no validation step.
-template <typename InstT, typename TransformFn, typename... EachFieldIdT>
-static auto TransformIfFieldsAreConstant(EvalContext& eval_context,
-                                         SemIR::Inst inst,
-                                         TransformFn transform_fn,
-                                         EachFieldIdT InstT::*... each_field_id)
-    -> SemIR::ConstantId {
-  return RebuildIfFieldsAreConstantImpl(
-      eval_context, inst, [](...) { return true; }, transform_fn,
-      each_field_id...);
+// Given the stored value `arg` of an instruction field and its corresponding
+// kind `kind`, returns the constant value to use for that field, if it has a
+// constant phase. `*phase` is updated to include the new constant value. If
+// the resulting phase is not constant, the returned value is not useful and
+// will typically be `InvalidIndex`.
```

Comment:
I thought we don't have `InvalidIndex` anymore?
```suggestion
// will typically be `NoneIndex`.
```